Create action pipeline for automatic build and test on push

MissPeppermintPatty opened the pull request:
[why]
We need a mechanism to automatically build and test the library with a GitHub Action pipeline informing the developer with its result.

[how]
As a first approach the required YAML files are added to the repro to allow to build & test the library for the latest stable Ubuntu release (currently Ubuntu 20.04). Further targets can be added with future commits. The action is automatically triggered each time a developer is pushing its sources to GitHub's taskolib.

Finii reviewed (changes requested):
Several questions

*    Why a 'dispatcher' workflow `ci.yml` at all?
*    Why on every push? Usually you do it on pull-request and/or push to `main`
*    Why a cron job, as it will be done on any push anyhow, the run will be identical?!
*    Why cache Meson at all, caching makes usually only sense if you BUILD something, but fetching the tar or fetching the cache is virtually the same
*    Download Meson artifact ... use ready action instead of writing code?
*    Sometimes `run` is combined with `shell`, sometimes not. Usually you do not need it
*    Why two Steps for 2 consecutive lines of output (`build_and_run_tests.yml` L 18 + 19), should that not be 1 step with two output lines?
*    `apt-get` and `wget`, why not `apt` and `curl` nowadays? (About the `wget`ting see note above)

Sorry ;-)

Finii commented:
Further ideas:

* I often add a `workflow_dispatch` trigger to manually trigger a run
* Instead if the dispatcher workflow you can combine the parallel builds easily with a `strategy: matrix: ...` definition IN your build.yml
* Generally I have the feeling the names are rather long-ish
* Add list of files that can be ignored (or the other way around) (`push: path-ignore:`)

MissPeppermintPatty commented:
First I must say that I am new to the subject: I don't have any experiences with "action" items to build & tests the sources in GitHub.

As I reference I took the pipeline action files from [GUL14](https://github.com/gul-cpp/gul14/tree/main/.github/workflows) and remove unnecessary items from it (for example build for MacOS).

At first I wanted to run it smoothly and then reduce it to their minimal requirements

Now to your questions:

> - Why a 'dispatcher' workflow ci.yml at all?
Sure, we can simplify it and use only the `build_and_run_tests.yml`.

> - Why on every push? Usually you do it on pull-request and/or push to main
This means some tweaky changes... Suggestions?

```yaml
on:
   [push]
```

> - Why a cron job, as it will be done on any push anyhow, the run will be identical?!
Only for testing. I removed it 😉 

> - Why cache Meson at all, caching makes usually only sense if you BUILD something, but fetching the tar or fetching the cache is virtually the same
I took as a reference GUL14. If you have any other scenario then tell me... 😄 

> - Download Meson artifact ... use ready action instead of writing code?
I don't get it. Can you explain?

> - Sometimes run is combined with shell, sometimes not. Usually you do not need it
> - Why two Steps for 2 consecutive lines of output (build_and_run_tests.yml L 18 + 19), should that not be 1 step with two output lines?
Hmm, here I don't have any opinion...

> - apt-get and wget, why not apt and curl nowadays? (About the wgetting see note above)
Yep! I would also suggest that this is the modern way to download the sources. Will adapt it.

MissPeppermintPatty commented:
The taskomat library depends on GUL14. When meson is executed I tries to find libgul14 which fails:

```meson
...
Run-time dependency libgul14 found: NO (tried pkgconfig and cmake)
[20](https://github.com/taskolib/taskolib/actions/runs/3319354453/jobs/5484408953#step:9:21)

[21](https://github.com/taskolib/taskolib/actions/runs/3319354453/jobs/5484408953#step:9:22)
meson.build:136:0: ERROR: Dependency "libgul14" not found, tried pkgconfig and cmake
```

How do I solve the problem? My suggestion is download GUL14 from GitHub and compile it. This means a new `meson.build` file? Any suggestions?

Finii commented:
> This means some tweaky changes... Suggestions?

```
on:
  push:
    branches: [ main ]
    path-ignore:
      - 'README'
      - ...
  pull_request:
    branches: [ main ]
    path-ignore:
      ...
  workflow_dispatch:

```

Finii commented:
`workflow_dispatch` adds a 'run manually' button somewhere where the arrow points

![image](https://user-images.githubusercontent.com/16012374/197747222-41e1e8b2-4390-474a-a53e-40ecafe2346d.png)

Finii commented:
> How do I solve the problem? My suggestion is download GUL14 from GitHub and compile it. This means a new `meson.build` file? Any suggestions?

No, that should be working when `wrap` has been set up correctly, which obviously isnt, I will fix.

Finii commented:
> Download Meson artifact ... use ready action instead of writing code?

```
steps:
  - name: Fetch ...
  uses: robinraju/release-downloader@v1.5
  with:
    repository: "mesonbuild/meson"
    tag: "0.63.0"
    fileName: "*.tar.gz"
 ```

This has the advantage that you say more abstractedly what you want. And you can give globs for the file name, which is helpful for releases-to-fetch when the version tag increases (automatically) and the release file name changing, which most repos do after some years ;)

> Hmm, here I don't have any opinion...

Well, the final output is more reasable imho:

```diff
-    steps:
-    - run: echo "Job triggered by ${{ github.event_name }} event, running on ${{ runner.os }}."
-    - run: echo "Branch ${{ github.ref }} on repository ${{ github.repository }}"
+    steps:
+    - run: |
+      - echo "Job triggered by ${{ github.event_name }} event, running on ${{ runner.os }}."
+      - echo "Branch ${{ github.ref }} on repository ${{ github.repository }}"
```

Finii commented:
Some tests fail, but it compiles ;-)

![image](https://user-images.githubusercontent.com/16012374/197758658-778ed620-f4e3-446b-aacc-c8715425e157.png)

Finii commented:
I guess it is not really helpfull to store _everything_ ...

![image](https://user-images.githubusercontent.com/16012374/197760615-6b5a0b11-2430-4723-9d9c-e4791e0f5408.png)

I know this has been set up by Lars for GUL, but imho this does not help and just makes downloading the error logs a longer task.

Here is what went wrong in the tests:

```
==================================== 1/2 =====================================
test:         taskomat / all
start time:   11:11:37
duration:     0.68s
result:       exit status 1
command:      MALLOC_PERTURB_=40 /home/runner/work/taskolib/taskolib/build.asan/taskomat_test
----------------------------------- stdout -----------------------------------

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
taskomat_test is a Catch v2.13.7 host application.
Run with -? for options

-------------------------------------------------------------------------------
Executor: Run a failing sequence asynchronously
-------------------------------------------------------------------------------
../tests/test_Executor.cc:122
...............................................................................

../tests/test_Executor.cc:143: FAILED:
  REQUIRE( executor.is_busy() == true )
with expansion:
  false == true

ERROR: Script execution error: sol: runtime error: [string "if a == 4 then terminate_sequence() end"]:1: 🛑ABORT🛑🛑ABORT🛑
stack traceback:
	[C]: in function 'base.terminate_sequence'
	[string "if a == 4 then terminate_sequence() end"]:1: in main chunk
===============================================================================
test cases:   193 |   192 passed | 1 failed
assertions: 12111 | 12110 passed | 1 failed
```

I guess ... OF COURSE you can not check if some other thread is "still running" because you do not know how much time passed meanwhile (on a heavy used build machine) ;) But that is another Issue and has nothing to do with this CI PR :tada:

alt-graph commented:
I am a little confused: This is marked as "ready", but the title still says "draft". Should we review it or not?

Finii commented:
> Should we review it or not?

I did review and requested changes, thereafter change to draft? iirc.

MissPeppermintPatty commented:
Sorry! Removed 'Draft' but I will push it on Monday ... too much other stuff to solve 😶

MissPeppermintPatty reviewed (commented):
I'm not familiar with 'wrapping' in meson. I would suppose here is the docu side: https://mesonbuild.com/Wrap-dependency-system-manual.html

BTW: I cannot approve the changes (it is disabled) - why?

Finii commented:
> I'm not familiar with 'wrapping' in meson. I would suppose here is the docu side: https://mesonbuild.com/Wrap-dependency-system-manual.html

We can remove that commit, already in `main` with latest Meson related PR.

I guess the starting point in Mesons docu is rather https://mesonbuild.com/Subprojects.html
The Subprojects then use wraps (i.e. a description how/where to get that subproject from)

> BTW: I cannot approve the changes (it is disabled) - why?

The changes were requested by my review, I can approve the changes.

But, I see no commit after my review, so nothing has been addressed?

Finii commented:
> ```yaml
> on:
>    [push]
> ```
> 
> > * Why a cron job, as it will be done on any push anyhow, the run will be identical?!
> 
> Only for testing. I removed it wink


![image](https://user-images.githubusercontent.com/16012374/199934027-8f637be1-919a-4656-81d4-8e8e38e65933.png)

No?

Finii commented on meson.build:
```diff
@@ -133,7 +133,7 @@ install_headers(files(
 inc = [ include_directories('include', 'include/taskomat/sol', 'src') ]
 
-gul_dep = dependency('libgul14')
+gul_dep = dependency('libgul14', version : '> 2.6', fallback : ['libgul14', 'libgul_dep'])
```
Ah, this wrap was better than the one we (I) pulled to `main` (i.e. contains version).
So maybe keep this commit.

MissPeppermintPatty replied:
Ok.

MissPeppermintPatty commented:
@Fini I saw that by using meson 'wrap' (`subprojects/libgul14.wrap`) we have now a new build folder `subprojects/libgul14`. I would added it to `.gitignore` to avoid pollution of the repository.

Finii commented:
> @fini I saw that by using meson 'wrap' (`subprojects/libgul14.wrap`) we have now a new build folder `subprojects/libgul14`. I would added it to `.gitignore` to avoid pollution of the repository.

Oh, yes, absolutely. It seems the gitignore is almost empty?

Put in `subprojects/*/`, and `build*/`, and some VScode stuff?

alt-graph commented:
> Elsewhere we have additionally (which I have no clue is needed for what, maybe Lars knows):
> 
> ```
> ._*
> nbproject
> out
> .vs
> obj
> vcpkg_installed
> ```

The only thing I would copy from this list is `._*`. On non-native Mac filesystems, Apple boxes generate files with this prefix to store metadata for other files.

fini commented:
Just a heads up, you mentioned the wrong fini :) Was a bit confused until I solved the mystery of why I was mentioned here 🕵️‍♂️

MissPeppermintPatty commented:
Oh oh... I try to rebase & merge my local sources and pushed it to the repository. Unfortunately we have now the changes which you already have committed (library name changed etc.).  Here I am lost ... should I merge or better start from scratch ... 🤔

Finii commented:
Just go back ... `git reflog` ... and force push etc, then you are where you began.

Finii commented:
I mean ... git never forgets anyting (except when it does). Whatever you do with git, you can always go back. Always. `git reflog` shows you all the recent things you did and shows you the hash to go back to exactly THAT state. `git reset --hard <hash>`.

Of course only things that had a commit, if pushed or not is not important.

Finii commented:
Example output in my local repo doing `git reflog`:

![image](https://user-images.githubusercontent.com/16012374/200011246-17b5ee97-a7d4-4f71-ad4e-f9abc92c5052.png)

Assume the `git rebase` went awry, ... just return to the previous state (hash) with `reset --hard`.

Finii commented:
You see, you can also undo `git commit --amend` for example, which adds stuff to an exiting commit. Sometimes afterwards you see that you amended the wrong commit.

MissPeppermintPatty commented:
> Just a heads up, you mentioned the wrong fini :) Was a bit confused until I solved the mystery of why I was mentioned here 🕵️‍♂️

@fini Argh ... sorry! But with the tiny little mixing in the first letter on your 'nickname' I pushed the comment to the wrong developer .. apologise me 🖖

MissPeppermintPatty commented:
Ok, thanks for all the comments but I will continue on Monday! ... I somehow feel like my head is exploding 🤯 🤪 😵 ...

MissPeppermintPatty reviewed (commented):
With the help of Lars I solved the problem with the somehow 'corrupted' sources.

alt-graph reviewed (approved):
Looks good to me – but I am not really experienced with Github actions, Fini is the local guru. :)

MissPeppermintPatty commented:
The main approach is to have a running CI pipeline. However I do not want to focus to extensively on __how__ it is implemented. As I said: I am a novice to GitHub actions and just learning...

> Why a 'dispatcher' workflow ci.yml at all?

As a first approach I would leave it as it is. Later we can disentangle it and bring it to a cleaner structure (in another PR)

> Why on every push? Usually you do it on pull-request and/or push to main

Fixed.

> Why a cron job, as it will be done on any push anyhow, the run will be identical?!

Removed.

> Why cache Meson at all, caching makes usually only sense if you BUILD something, but fetching the tar or fetching the cache is virtually the same
> Download Meson artifact ... use ready action instead of writing code?

```
- name: Fetch ...
  uses: robinraju/release-downloader@v1.5
  with:
    repository: "mesonbuild/meson"
    tag: "0.63.0"
    fileName: "*.tar.gz"
```
Your approach simplifies the action and I pushed your changes to the repo (... also changed to the newest release `...@v1.6` 😉 ).  By the way, do you know how to trigger the process [manually](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow)? I read that using [workflow_dispatch](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#workflow_dispatch) is a solution but unfortunately it did not appear in the GitHub webpages; I add `workflow_dispatch` to [ci.yml](https://github.com/taskolib/taskolib/pull/17/commits/76a18c11e52caf8e06e64fcfb7e7142c9bd1f0a8#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03f). The reference of [Doxygen update](https://github.com/taskolib/taskolib/actions/workflows/doxygen.yml) has the `Run workflow` button. Background is that I wanted to test the modified changes also manually.
![Screenshot 2022-11-08 at 18 47 13](https://user-images.githubusercontent.com/53872500/200638246-f03b7367-ec26-4056-92dc-688ee7cd46df.png)

> Sometimes run is combined with shell, sometimes not. Usually you do not need it

Fixed

> Why two Steps for 2 consecutive lines of output (build_and_run_tests.yml L 18 + 19), should that not be 1 step with two output lines?

Fixed

> apt-get and wget, why not apt and curl nowadays? (About the wgetting see note above)

Now we use `apt` instead of the old fashioned `apt-get`. With the above approach for installing meson `curl` is not needed anymore. But we should keep in mind that we must use the [-L option](https://curl.se/docs/manpage.html) to download meson because the http link is a redirected URL.

Then I wanted to answer to your [request in the conversation](https://github.com/taskolib/taskolib/pull/17#pullrequestreview-1154378909) but I have now clue on how to handle a reply?  
![Screenshot 2022-11-08 at 18 44 20](https://user-images.githubusercontent.com/53872500/200637698-ae8e14ca-bffc-4cb1-95f9-ebbe5fd89e18.png)

Finii commented:
> > Why a 'dispatcher' workflow ci.yml at all?
> 
> As a first approach I would leave it as it is. Later we can disentangle it and bring it to a cleaner structure (in another PR)

I would really remove it. This is not how things are done on github, see comment above. Except there are reasons to have this (which are rather expert expert stuff).

On gitLAB, as far as I know, you need that approach also for matrix setups, with gitHUB there is the `strategy: matrix` to spawn a lot of different sub-processes that works very nice and without strange extra files.
In fact here we would set up the `platform` in a `matrix` and then have the process run in parallel on all platforms we want.

> By the way, do you know how to trigger the process manually
> [...] but unfortunately it did not appear in the GitHub webpages

It will only appear once the workflow is INSTALLED (I mean is in `main`). You can not interact nicely with workflows that are in some branch. It will work once it is pulled.

> Background is that I wanted to test the modified changes also manually.

The usual solution is to fork the repo and push the changed workflows to `main` on your fork.
That is a bit more work, but the only way to test.

> > Sometimes run is combined with shell, sometimes not. Usually you do not need it
> 
> Fixed

![image](https://user-images.githubusercontent.com/16012374/200642827-d95a2e5b-003b-4a9c-870d-4d0cffa429e3.png)

> Then I wanted to answer to your but I have now clue on how to handle a reply?

That is unfortunately not possible.
Also not possible is `suggestion` in reviews that have multiple lines. These two features are clearly superior on gitLAB. Also the whole 'approve' cycle is a bit ... cumbersome on gitHUB imho. /shrug

Finii commented:
> It will only appear once the workflow is INSTALLED (I mean is in `main`). You can not interact nicely with workflows that are in some branch. It will work once it is pulled.
> 
> > Background is that I wanted to test the modified changes also manually.
> 
> The usual solution is to fork the repo and push the changed workflows to `main` on your fork. That is a bit more work, but the only way to test.

Well, the **reason** is safety. The CI run has some rights and runs under a user, and usually CI files in not-the-default-branch are un-approved by the Maintainers. Everyone can create them. If you can trigger them it is very hard to decide which rights they should run under. There is a bit of change at the moment how the rights management for CI jobs is handled at github, but ... anyhow.
For example a CI job can create a docker container of my project and push it to dockerhub.
Now you, a malicious developer, create a PR, that changes the workflow and pushes something else to 'our' docker account. This would not be good? So only `main`, which is usually protected and where only 'certified persons' can pull to, is fully functional.

Finii commented:
Well, maybe off-topic, this is one of my docker CIs ;-)

```yml
name: Docker release

on:
  push:
    branches:
      - master
    paths:
      # Keep this in line with .dockerignore
      - .....
      - .dockerignore
      - Dockerfile
  workflow_dispatch:

jobs:
  publish-image:
    name: Publish image
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v3

        # Docker Meta provides automatic, standard tagging
      - name: Docker meta
        id: meta
        uses: docker/metadata-action@v4
        with:
          images: mydocker/account
          flavor: latest=true

        # QEMU allows us to build container images for multiple architectures (amd64, ARM etc...)
      - name: Set up QEMU
        uses: docker/setup-qemu-action@v2
        with:
          platforms: linux/amd64,linux/arm64

        # Docker Buildx provides Buildkit support
      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v2
        with:
          platforms: linux/amd64,linux/arm64

      - name: Login to Docker Hub
        uses: docker/login-action@v2
        with:
          username: ${{ secrets.DOCKER_USER }}
          password: ${{ secrets.DOCKER_PAT }}

      - name: Build and push
        uses: docker/build-push-action@v3
        with:
          context: .
          platforms: linux/amd64,linux/arm64
          push: true
          tags:  ${{ steps.meta.outputs.tags }}
          labels: ${{ steps.meta.outputs.labels }}
```

And you see the docker account password is in the `secrets`. If any branch could run and use them this is a security breach.

This is the same with my `MCS-gitlab` pulls `GUL` from github, my pass-token is hidden in the `secrets` and could be mis-used if everyone was allowed to change the workflow in some branch (well, pipeline they call it).

Finii commented:
> I somehow feel like my head is exploding :exploding_head: :zany_face: :dizzy_face: ...

Great job, anyhow :tada:

Finii reviewed (changes requested):
This looks much better :+1:

Here are some suggestions
* The job ID must be alphanumeric
* Instead add the text as name
* Not sure why you tried to specify the platform in the way you did, but that does not work
* Drop platform test for Ninja (we know the platform is Ubuntu)
* Simplify switching Meson version
* Switch to Meson version 'we' have on Ubuntu 20.04

```diff
diff --git a/.github/workflows/build_and_run_tests.yml b/.github/workflows/build_and_run_tests.yml
index 7e47c53..cf8d5d0 100644
--- a/.github/workflows/build_and_run_tests.yml
+++ b/.github/workflows/build_and_run_tests.yml
@@ -25,8 +25,9 @@ on:
 
 # A workflow run is made up of one or more jobs that can run sequentially or in parallel
 jobs:
-  build-and-run-tests-on-ubuntu-20.04:
-    runs-on: ${{ inputs.platform }}
+  buildntest:
+    name: "Build and Tests (Ubuntu-20.04)"
+    runs-on: ubuntu-20.04
     steps:
     - run: |
         echo "Job triggered by ${{ github.event_name }} event, running on ${{ runner.os }}."
@@ -37,28 +38,23 @@ jobs:
 
     - name: Install Ninja
       run: |
-        if [[ "${{ inputs.platform }}" == ubuntu-latest ]]; then
-          sudo apt install ninja-build
-        else
-          echo "Do not know how to install Ninja on platform ${{ inputs.platform }}"
-          false
-        fi
+        sudo apt install ninja-build
 
-    - name: Fetch meson version 0.63.0 for building
+    - name: Fetch newer meson version for building
       uses: robinraju/release-downloader@v1.6
       with:
         repository: "mesonbuild/meson"
-        tag: "0.63.0"
+        tag: "0.53.2"
         fileName: "*.tar.gz"
 
     - name: Unpack meson tar ball and remove downloaded file
       run: |
-        tar -xzf meson-0.63.0.tar.gz
-        rm meson-0.63.0.tar.gz
+        tar -xzf meson-*.tar.gz
+        rm meson-*.tar.gz
 
     - name: Create symbolic link for Meson
       run: |
-        ln -s meson-0.63.0/meson.py .
+        ln -s meson-*/meson.py .
 
     - name: Configure build directories
       run: |
@@ -70,5 +66,5 @@ jobs:
 
     - name: Build and run ASAN tests
       run: ./meson.py test -C build.asan
-  with:
-    platforms: ubuntu-20.04
```

Finii commented:
Ah, and if you want to run it on multiple different platforms, use a matrix maybe like this:

```diff
 # A workflow run is made up of one or more jobs that can run sequentially or in parallel
 jobs:
   buildntest:
-    name: "Build and Tests (Ubuntu-20.04)"
-    runs-on: ubuntu-20.04 #${{ inputs.platform }}
+    name: "Build and Tests"
+    strategy:
+      matrix:
+        os: [ ubuntu-20.04, ubuntu-22.04 ]
+    runs-on: ${{ matrix.os }}
+    #runs-on: ubuntu-20.04 #${{ inputs.platform }}
     steps:
     - run: |
         echo "Job triggered by ${{ github.event_name }} event, running on ${{ runner.os }}."
```

The concrete matrix variable for that one job will be added to the job's name automatically:

![image](https://user-images.githubusercontent.com/16012374/201281779-350c6632-1139-4284-b72f-b941b54e90a5.png)

MissPeppermintPatty commented:
### Forking
> It will only appear once the workflow is INSTALLED (I mean is in main). You can not interact nicely with workflows that are in some branch. It will work once it is pulled.
>
> > Background is that I wanted to test the modified changes also manually.
>
> The usual solution is to fork the repo and push the changed workflows to main on your fork. That is a bit more work, but the only way to test.

I have some question about forking.

[1] Here I found your fork to improve the CI workflow... Thanks again so much for supporting me!
[2] I am the only maintainer of the fork (read/write), but other GitHub contributers have read access to the repo?
[3] Can I give it beside 'taskolib' also any other free text name, for example `TheAmazingSequencerLibrary`?
[4] When disabling `main branch only` GitHub will fork also other branches? Is there a dynamic sync mechanism during live-time of the repo or does it only creates a fork of the actual branches and than decouples it from the origin repo?
[5] Selecting "taskolib" as `Owner` any contributors of the taskolib can read and write to the folk? What about other GitHub contributors? They only have read access?

![Screenshot 2022-11-11 at 09 36 22](https://user-images.githubusercontent.com/53872500/201304211-be6d0110-ff68-4932-b2ae-ccf91caafb2c.png)

However I found a good practical hands-on webpage: [Chaser324: GitHub Forking](https://gist.github.com/Chaser324/ce0505fbed06b947d962).

### 'API' document
Do you have any webpage showing the 'API' of GitHub actions? Not the REST API. As far as I can see it on the web we have only a 'prose' text format but not a structure document most likely based on something like [Augmented Backus-Naur Form](https://en.wikipedia.org/wiki/Augmented_Backus–Naur_form) 🤔 ?

Finii commented:
> [1] Here I found your fork to improve the CI workflow... Thanks again so much for supporting me!

The more you learn the less I have to do in the long run :-D

> [2] I am the only maintainer of the fork (read/write), but other GitHub contributers have read access to the repo?

You can change the visibility of your repo (regardless if it started as fork or newly created). Well, private repos need 'pro'.
If it is a fork people seeing your repo will also see that it is a fork (if you forked via Github and not manually).

![image](https://user-images.githubusercontent.com/16012374/201318637-31ffc435-66c2-496b-9851-215325a53dd5.png)

> [3] Can I give it beside 'taskolib' also any other free text name, for example `TheAmazingSequencerLibrary`?

Why not?

> [4] When disabling `main branch only` GitHub will fork also other branches?

Yes. For examle the `gh-pages` and so on. In former times Github always did a complete fork, but as people often just want to submit a PR to the original repo and do not need all the branches, they now only include one branch per default, to keep the fork's repo size small(er). The fork is a complete copy and talkes as much disk-spaces as the original, the commit blobs are NOT shared. Some repos are forked thousand of times...

> Is there a dynamic sync mechanism during live-time of the repo or does it only creates a fork of the actual branches and than decouples it from the origin repo?

A fork is just 'initiate a copy now'.
But you can sync very easily via workflow :-D

> [5] Selecting "taskolib" as `Owner` any contributors of the taskolib can read and write to the folk?

That means all members of the organization have rw access; contributors of `taskolib` is something different from `member of taskolib orga`.

> What about other GitHub contributors? They only have read access?

Yes. Unless you invite them to become either a contributor or a member of the organization that the repo is under.

> Do you have any webpage showing the 'API' of GitHub actions? Not the REST API. As far as I can see it on the web we have only a 'prose' text format but not a structure document most likely based on something like [Augmented Backus-Naur Form](https://en.wikipedia.org/wiki/Augmented_Backus%E2%80%93Naur_form) thinking ?

You mean the DSL that is used in the `ci.yml` files? Well, it is no language although it looks like one. It is just a yaml data block.

The allowed/needed members are - as far as I know - all just prose.

Finii commented:
> > [3] Can I give it beside 'taskolib' also any other free text name, for example `TheAmazingSequencerLibrary`?
> 
> Why not?

![image](https://user-images.githubusercontent.com/16012374/201320792-08e4c984-b6ee-4d70-a6b5-12b699622abd.png)

Hehe :smiling_imp:

Finii commented:
> However I found a good practical hands-on webpage: [Chaser324: GitHub Forking](https://gist.github.com/Chaser324/ce0505fbed06b947d962).

I usually point to https://egghead.io/courses/how-to-contribute-to-an-open-source-project-on-github

Well, the how to sync instructions from Chaser324 are a bit ... simple. That does not really sync; for example TAGs are missing.

I do it btw most often the other way around. `origin` is the original and I call my fork `fork` in the git remotes. In former times I did it as he says, but at some point I always became confused if 'origin' is the original or the fork, and then I shifted. (Especially if you want to push to both).
Keep in mind that `pull` and `fetch` only work on one remote at a time and often you want to also `fetch fork` or something.
This is also part of the reason that I said (in the mail) 'do `git fetch` and not `git pull`', because maybe you want to fetch all remotes and then sort out what to do.

Finii commented:
> I do it btw most often the other way around. `origin` is the original and I call my fork `fork` in the git remotes.

This is because I usually clone the original first. Like you in your current situation. You already worked on the original repo, and you have a clone locally.
Now you fork, because suddenly you realize you need a fork to test stuff or something. You fork on github and then add the fork to the old/current repo on you machine with `git remote add fork <fork's URI>`.

Matter of taste. But I have been bitten a lot by the suggested 'upstream' approach. For example on the doocsdev20 machine I work on GUL and checked GUL out normally (origin == original). On my desktop machine I did the dance they suggest, because I want to debug the workflows: Fork, clone the fork (origin == fork) and add the original (upstream == original).

Now I work sometimes on doocsdev20 and sometimes on my local desktop machine, and what `origin` is is different on both. Not good.
And because forks come later in the process, I almost ALWAYS clone the original first, even for projects I did not join yet, etc, I finally shifted to approach such that `origin` == original.

Finii commented:
Sorry for spamming ... I think that point of naming IS important.

Lets see how it was when I tried to help you with this commit...

```
git clone https://github.com/taskolib/taskolib.git
... worked a lot with the repo, `origin` = original
git fetch
git checkout feature/add_ci_pipeline
... checks out your PR branch, create new local branch, switchs to it etc
... work on it
... need to debug workflows
... ... create fork on github
git remote add fork https://github.com/Finii/taskolib.git
git fetch fork
... push edited stuff and try it out
git push --force fork HEAD:fork/main
```

Note the last command. Everything is obvious and spelled out this way.
The local branch *still* follows the original, and I can do a `git push` once I'm certain it works (on my fork).

MissPeppermintPatty commented:
@Finii Finally it works, also regarding to your suggestions! 

### Failure and the Aha-effect
By running the action a first time I got the following strange error:
![Screenshot 2022-11-11 at 12 39 25](https://user-images.githubusercontent.com/53872500/201363114-af92aca5-f79b-4819-b498-c4670611b75b.png)

[1]: Why at line 1? The comment is completely ok. Mysterious ... But then ...
[2]: ... I saw that GitHub claims about the `with` at the end of the `steps`. Then I realise that the item `jobs` doesn't have any `with` in this context. The interpreter was expecting a [jobs.<job_id>.uses](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_iduses) and the [jobs.<job_id>.with](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idwithinput_id) -> GitHub claims it as a syntax error. And as you already mentioned [jobs.<job_id>.with.<input_id>](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idwithinput_id). Also the webpages [input context](https://docs.github.com/en/actions/learn-github-actions/contexts#inputs-context) gives more insights. My foolishness 🙁 . 

### Input 'parameters'
About the issue of `${{input.platform}}` in the _Install Ninja_ block. It became clear to me that the phrase `input` means that we have external parameters that are transferred to the 'build script'. After looking at the 'plugin' [robinraju/release-downloader](https://github.com/robinraju/release-downloader) it gives a good insight on how it works. Also in conjunction with the preceding note. I would claim it is some _interfacing_ of external parameters.

### Matrix build
The `matrix` will simplify the build process on different platforms. Thanks for that 😄 . 

### Duplication
I wanted to replace the duplicated `path-ignore` items but until today I do not find any solution for that. Needs deeper investigation...  

### Meson 0.53.2
Why do you install the previous release 0.53.2 of meson? Because to sync it with our local build hosts?

### Fork
Actually using a fork helps much to test it 😉. I will try to push your comments above about forking to my mind...

Finii commented:
### Input parameters

Well, you can create variables of various types along the one.
`input.` is for when you call this from another flow
But you can also put something into the environment of the shell (for the next steps), or you can directly define values that a step sets, for example

```
jobs:
  one_job:
    runs-on: ubuntu-latest
    steps:

    - name 'a step'
    - id: most_important
    - run: |
      echo "level=100" >> $GITHUB_OUTPUT

    - name 'sometimes run'
    - if: ${{ steps.most_important.outputs.level > 10 }}
    - run: |
      echo Jo!
      echo "ANSWER=JA" >> $GITHUB_ENV

    outputs:
      something: "Level ${{ steps.most_important.outputs.level }}"
      lololo: "Some verbatim stuff"

  other_job:
    runs-on: ubuntu-latest
    needs: one_job
    env:
      BLAH: ${{ needs.one_job.outputs.lololo }}
    steps:
    - name 'do it or not'
    - if: needs.one_job.outputs.something == "JA"
    - run: |
      echo "level=100" >> $GITHUB_OUTPUT
    
```

Write to the environment by echoing to GITHUB_ENV in the same manner.

Workflows can also have outputs, that can be set. Sometimes it is not so clear which way to transfer values is the best one...
You see some examples above (untested).

Ah, the `${{ }}` syntax is like bash substitution. You can omit that in `if:`s.

Just as ideas.

### De-duplication

I also did not manage to do it.

`on: [ push, pull_request ]` does work, but does not allow to specify filters
Variables seem to be not evaluated there (i.e. `path-ignore: ${{ something }}`)

The other question is if it is really needed to not-run if only files in that paths are changed. Usually that is only important if you ACT on the trigger, like create an additional commit in the repo or trigger a package release.

### Meson 0.53.2

Yes. I would not use a too new version, or we will not catch 'invalid' use of new Meson features. We guarantee at the moment that it runs with Meson >= 0.49 iirc, so maybe we should check that.

Of course you can set up a full matrix (that is what matrices are really about), check on 20.04 with meson 0.49, 0.53 and 0.63, and on 22.04 with all the three mesons, and on 22.10 ;-) The matrix will start jobs for all permutations.
(Please dont)

I have no clue, going to 53.2 was like ... do not use something too new if you guarantee that it works with something rather old. /shrug
